### ISSUE TITLE
[flowbroker] remote node persistence

### ISSUE THREAD

**Issue by betefaber** [dojot/dojot#1803]
Affected baseline **v0.4.4-rc.1**, kubernetes environment

- Add remote node
```
curl -X POST ${DOJOT_HOST}/flows/v1/node -H "Authorization: Bearer ${JWT}" -H "content-type: application/json" -d '{"image": "dojot/kelvin-example:3.0.0-alpha2","id":"kelvin"}'
{"message":"ok"}
```
- Restart flowbroker
```
kubectl delete pod/flownode-adminkelvin-77659695-75b82
pod "flownode-adminkelvin-77659695-75b82" deleted
```
- apparently the node is added
```
<18:21:40 22/10/2020> -- |kb8sRemoveNode -- DEBUG: Building deployment creation request...
<18:21:40 22/10/2020> -- |kb8sRemoveNode -- DEBUG: Adding container dojot/kelvin-example:3.0.0-alpha2 to the set...
<18:21:40 22/10/2020> -- |kb8sRemoveNode -- DEBUG: ... container adminkelvin was added to the set.
<18:21:40 22/10/2020> -- |kb8sRemoveNode -- DEBUG: ... deployment creation request was built.
<18:21:40 22/10/2020> -- |kb8sRemoveNode -- DEBUG: Deployment is:
<18:21:40 22/10/2020> -- |kb8sRemoveNode -- DEBUG: { apiVersion: 'apps/v1',
  kind: 'Deployment',
  metadata:
   { labels: { app: 'flowbroker', name: 'flownode-adminkelvin' },
     name: 'flownode-adminkelvin' },
  spec:
   { selector:
      { matchLabels: { app: 'flowbroker', name: 'flownode-adminkelvin' } },
     replicas: 1,
     template:
      { metadata: { labels: { name: 'flownode-adminkelvin', app: 'flowbroker' } },
        spec:
         { containers:
            [ { name: 'adminkelvin',
                image: 'dojot/kelvin-example:3.0.0-alpha2',
                imagePullPolicy: 'Always',
                ports: [ { name: 'amqp', port: 5555, containerPort: 5555 } ] } ],
           restartPolicy: 'Always' } } } }
<18:21:40 22/10/2020> -- |dispatcher -- INFO: Dispatcher connected to undefined on port 5555
<18:21:40 22/10/2020> -- |kb8sRemoveNode -- DEBUG: Current flowbroker deployments are: { statu
...
<18:21:40 22/10/2020> -- |kb8sRemoveNode -- DEBUG: ... server access is OK.
```

- but the node is not persisted

![image](https://user-images.githubusercontent.com/37310063/96919246-e168a380-1481-11eb-8630-ef9cd9595bc6.png)

- although the remote node exists, it cannot be removed
```
curl -X GET "${DOJOT_HOST}/flows/v1/node" -H "Authorization: Bearer ${JWT}" -H "Content-Type:application/json"
{"nodes":[{"id":"kelvin","image":"dojot/kelvin-example:3.0.0-alpha2"}]}
```
```
curl -X DELETE ${DOJOT_HOST}/flows/v1/node/kelvin -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{"message":"Failed to remove node.","error":"Not found remote node with id kelvin."}
```
```

<18:34:07 22/10/2020> -- |api -- ERROR: Error: Not found remote node with id kelvin.
    at Promise (/opt/flowbroker/orchestrator/src/nodeManager.js:349:25)
    at new Promise (<anonymous>)
    at NodeManager.delRemoteNode (/opt/flowbroker/orchestrator/src/nodeManager.js:318:12)
    at app.delete (/opt/flowbroker/orchestrator/src/api.js:78:15)
    at Layer.handle [as handle_request] (/opt/flowbroker/orchestrator/node_modules/express/lib/router/layer.js:95:5)
    at next (/opt/flowbroker/orchestrator/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/opt/flowbroker/orchestrator/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/opt/flowbroker/orchestrator/node_modules/express/lib/router/layer.js:95:5)
    at /opt/flowbroker/orchestrator/node_modules/express/lib/router/index.js:281:22
    at param (/opt/flowbroker/orchestrator/node_modules/express/lib/router/index.js:354:14)
    at param (/opt/flowbroker/orchestrator/node_modules/express/lib/router/index.js:365:14)
    at Function.process_params (/opt/flowbroker/orchestrator/node_modules/express/lib/router/index.js:410:3)
    at next (/opt/flowbroker/orchestrator/node_modules/express/lib/router/index.js:275:10)
    at serveStatic (/opt/flowbroker/orchestrator/node_modules/serve-static/index.js:75:16)
    at Layer.handle [as handle_request] (/opt/flowbroker/orchestrator/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/opt/flowbroker/orchestrator/node_modules/express/lib/router/index.js:317:13)
```
 
Note: related to #742

**Comment by bizetti** [comment]
0.4.4 rc2 e 0.5.0.rc2 (K8s)

**Comment by betefaber** [comment]
**Comparing between docker and k8s:**

----
docker environment
----
1. dojot up
2. add remote node
3. dojot down
4. dojot up: the remote node is restored
5. **restart flowbroker**: the node is persisted

```
flowbroker_1                  | <12:36:13 23/10/2020> -- |nodeMngr -- DEBUG: Succeeded to remove container.
flowbroker_1                  | <12:36:13 23/10/2020> -- |nodeMngr -- DEBUG: Succeeded to remove container.
flowbroker_1                  | <12:36:14 23/10/2020> -- DEBUG: Unsubscribing from topics ff2ed6c7-9bf2-482a-a430-f0c90be846b5,39419cc2-6780-4076-bbe1-383e72ce669a,1d32b569-fb58-488e-9f15-a3b1ec454c89,6d957d3c-b37a-43dd-be7e-cd88f47f6deb,3920abad-0476-4cc1-96c7-1f29d37cc902,ec94fd58-f3f6-4715-922d-ea9fe5d8e62a
flowbroker_1                  | <12:36:14 23/10/2020> -- DEBUG: Subscribing to topics ff2ed6c7-9bf2-482a-a430-f0c90be846b5,39419cc2-6780-4076-bbe1-383e72ce669a,1d32b569-fb58-488e-9f15-a3b1ec454c89,6d957d3c-b37a-43dd-be7e-cd88f47f6deb,3920abad-0476-4cc1-96c7-1f29d37cc902,ec94fd58-f3f6-4715-922d-ea9fe5d8e62a
flowbroker_1                  | [harbor-master][modem] error parsing body as json
flowbroker_1                  | <12:36:15 23/10/2020> -- |dockerRemoteNode -- DEBUG: [nodes] image dojot/kelvin-example:3.0.0-alpha2 created
flowbroker_1                  | [harbor-master][modem] error parsing body as json
flowbroker_1                  | <12:36:15 23/10/2020> -- |dockerRemoteNode -- DEBUG: [nodes] image dojot/kelvin-example:3.0.0-alpha2 created
flowbroker_1                  | <12:36:16 23/10/2020> -- |dockerRemoteNode -- DEBUG: [nodes] container flowbroker.adminkelvin.d90e2d1 was created
flowbroker_1                  | <12:36:16 23/10/2020> -- |dockerRemoteNode -- DEBUG: [nodes] container flowbroker.testekelvin.2cdae47 was created
flowbroker_1                  | <12:36:18 23/10/2020> -- |dockerRemoteNode -- DEBUG: Target: 4a6678d42a05
flowbroker_1                  | <12:36:18 23/10/2020> -- |dockerRemoteNode -- DEBUG: Target: 63ded8949152
flowbroker_1                  | <12:36:18 23/10/2020> -- |dockerRemoteNode -- DEBUG: [nodes] container up: flowbroker.adminkelvin.d90e2d1:4a6678d42a054ba11b954719fef8246de239cf0a5f1a00e86c86908c33cc08ff
flowbroker_1                  | <12:36:18 23/10/2020> -- |dispatcher -- INFO: Dispatcher connected to 4a6678d42a05 on port 5555
flowbroker_1                  | <12:36:18 23/10/2020> -- |dispatcher -- DEBUG: Received reply: {"payload":{"payload":{"id":"sample/kelvin","name":"kelvin","module":"kelvin","version":"0.0.0"}},"requestId":"2e630828-acdc-4486-b63d-5e44eac01fa8"}
flowbroker_1                  | <12:36:18 23/10/2020> -- |dispatcher -- DEBUG: Received reply: {"payload":{"payload":"<script type=\"text/x-red\" data-template-name=\"kelvin\">\n    <div class=\"form-row\">\n        <label for=\"node-input-name\"><i class=\"fa fa-tag\"></i> Name</label>\n        <input type=\"text\" id=\"node-input-name\">\n    </div>\n    <div class=\"form-row\">\n        <label for=\"node-input-in\"><i class=\"fa fa-edit\"></i> <span>input from</span></label>\n        <input type=\"text\" id=\"node-input-in\" placeholder=\"payload\" style=\"width:250px;\">\n        <input type=\"hidden\" id=\"node-input-inFieldType\">\n    </div>\n    <div class=\"form-row\">\n        <label for=\"node-input-out\"><i class=\"fa fa-edit\"></i> <span>output to</span></label>\n        <input type=\"text\" id=\"node-input-out\" placeholder=\"payload\" style=\"width:250px;\">\n        <input type=\"hidden\" id=\"node-input-outFieldType\">\n    </div>\n</script>\n\n<script type=\"text/x-red\" data-help-name=\"kelvin\">\n</script>\n\n<script type=\"text/javascript\">\n    RED.nodes.registerType('kelvin', {\n        category: 'converters',\n        color: \"rgb(231, 231, 174)\",\n        defaults: {\n            name: { value: \"\", required: false },\n            out:  { value: \"payload\", validate: RED.validators.typedInput(\"outFieldType\") },\n            outFieldType: { value: \"msg\" },\n            in: { value: \"payload\", validate: RED.validators.typedInput(\"inFieldType\") },\n            inFieldType: { value: \"msg\" },\n        },\n        inputs: 1,\n        outputs: 1,\n        icon: \"cog.png\",\n        label: function () {\n            return this.name || \"to kelvin\";\n        },\n        labelStyle: function () {\n            return this.name ? \"node_label_italic\" : \"\";\n        },\n        oneditprepare: function () {\n            if (!this.outFieldType) {\n                this.outFieldType = 'msg';\n            }\n\n            $(\"#node-input-out\").typedInput({\n                default: 'msg',\n                types: ['msg'],\n                typeField: $(\"#node-input-outFieldType\")\n            });\n\n            if (!this.inFieldType) {\n                this.inFieldType = 'msg';\n            }\n\n            $(\"#node-input-in\").typedInput({\n                default: 'msg',\n                types: ['msg'],\n                typeField: $(\"#node-input-inFieldType\")\n            });\n        },\n        oneditsave: function () { }\n    });\n</script>"},"requestId":"a3d9c6fd-ff80-4b52-8227-749eb13581e6"}
flowbroker_1                  | <12:36:20 23/10/2020> -- |dockerRemoteNode -- DEBUG: [nodes] container up: flowbroker.testekelvin.2cdae47:63ded8949152471c04bfc03202e027c233e0f0b0bbb20779cd5fa769ac6ac9a2
flowbroker_1                  | <12:36:20 23/10/2020> -- |dispatcher -- INFO: Dispatcher connected to 63ded8949152 on port 5555
flowbroker_1                  | <12:36:20 23/10/2020> -- |dispatcher -- DEBUG: Received reply: {"payload":{"payload":{"id":"sample/kelvin","name":"kelvin","module":"kelvin","version":"0.0.0"}},"requestId":"5ddf710c-5f2b-487f-87a8-ef52628bd66f"}
flowbroker_1                  | <12:36:20 23/10/2020> -- |dispatcher -- DEBUG: Received reply: {"payload":{"payload":"<script type=\"text/x-red\" data-template-name=\"kelvin\">\n    <div class=\"form-row\">\n        <label for=\"node-input-name\"><i class=\"fa fa-tag\"></i> Name</label>\n        <input type=\"text\" id=\"node-input-name\">\n    </div>\n    <div class=\"form-row\">\n        <label for=\"node-input-in\"><i class=\"fa fa-edit\"></i> <span>input from</span></label>\n        <input type=\"text\" id=\"node-input-in\" placeholder=\"payload\" style=\"width:250px;\">\n        <input type=\"hidden\" id=\"node-input-inFieldType\">\n    </div>\n    <div class=\"form-row\">\n        <label for=\"node-input-out\"><i class=\"fa fa-edit\"></i> <span>output to</span></label>\n        <input type=\"text\" id=\"node-input-out\" placeholder=\"payload\" style=\"width:250px;\">\n        <input type=\"hidden\" id=\"node-input-outFieldType\">\n    </div>\n</script>\n\n<script type=\"text/x-red\" data-help-name=\"kelvin\">\n</script>\n\n<script type=\"text/javascript\">\n    RED.nodes.registerType('kelvin', {\n        category: 'converters',\n        color: \"rgb(231, 231, 174)\",\n        defaults: {\n            name: { value: \"\", required: false },\n            out:  { value: \"payload\", validate: RED.validators.typedInput(\"outFieldType\") },\n            outFieldType: { value: \"msg\" },\n            in: { value: \"payload\", validate: RED.validators.typedInput(\"inFieldType\") },\n            inFieldType: { value: \"msg\" },\n        },\n        inputs: 1,\n        outputs: 1,\n        icon: \"cog.png\",\n        label: function () {\n            return this.name || \"to kelvin\";\n        },\n        labelStyle: function () {\n            return this.name ? \"node_label_italic\" : \"\";\n        },\n        oneditprepare: function () {\n            if (!this.outFieldType) {\n                this.outFieldType = 'msg';\n            }\n\n            $(\"#node-input-out\").typedInput({\n                default: 'msg',\n                types: ['msg'],\n                typeField: $(\"#node-input-outFieldType\")\n            });\n\n            if (!this.inFieldType) {\n                this.inFieldType = 'msg';\n            }\n\n            $(\"#node-input-in\").typedInput({\n                default: 'msg',\n                types: ['msg'],\n                typeField: $(\"#node-input-inFieldType\")\n            });\n        },\n        oneditsave: function () { }\n    });\n</script>"},"requestId":"95abd0c1-d131-4c16-9c9a-b3a32244a4ed"}
```

![Screenshot_61](https://user-images.githubusercontent.com/37310063/97017628-8936ad80-1524-11eb-8d92-d0733badc99f.png)

----
k8s environment:
----

1. dojot up

2. add remote node

![Screenshot_59](https://user-images.githubusercontent.com/37310063/97017704-a10e3180-1524-11eb-8d25-f742c89961b4.png)

3. **restart flowbroker**:
```
<14:16:12 23/10/2020> -- |kb8sRemoveNode -- DEBUG: Building deployment creation request...
<14:16:12 23/10/2020> -- |kb8sRemoveNode -- DEBUG: Adding container dojot/kelvin-example:3.0.0-alpha2 to the set...
<14:16:12 23/10/2020> -- |kb8sRemoveNode -- DEBUG: ... container adminkelvin was added to the set.
<14:16:12 23/10/2020> -- |kb8sRemoveNode -- DEBUG: ... deployment creation request was built.
<14:16:12 23/10/2020> -- |kb8sRemoveNode -- DEBUG: Deployment is:
<14:16:12 23/10/2020> -- |kb8sRemoveNode -- DEBUG: { apiVersion: 'apps/v1',
  kind: 'Deployment',
  metadata:
   { labels: { app: 'flowbroker', name: 'flownode-adminkelvin' },
     name: 'flownode-adminkelvin' },
  spec:
   { selector:
      { matchLabels: { app: 'flowbroker', name: 'flownode-adminkelvin' } },
     replicas: 1,
     template:
      { metadata: { labels: { name: 'flownode-adminkelvin', app: 'flowbroker' } },
        spec:
         { containers:
            [ { name: 'adminkelvin',
                image: 'dojot/kelvin-example:3.0.0-alpha2',
                imagePullPolicy: 'Always',
                ports: [ { name: 'amqp', port: 5555, containerPort: 5555 } ] } ],
           restartPolicy: 'Always' } } } }
<14:16:12 23/10/2020> -- |dispatcher -- INFO: Dispatcher connected to undefined on port 5555
<14:16:12 23/10/2020> -- |kb8sRemoveNode -- DEBUG: Current flowbroker deployments are: { statusCode: 200,
...
<14:16:12 23/10/2020> -- |kb8sRemoveNode -- DEBUG: ... server access is OK.
```
![Screenshot_58](https://user-images.githubusercontent.com/37310063/97017778-b5522e80-1524-11eb-82de-bf7994533868.png)

**Comment by betefaber** [comment]
Fixed in **v0.5.0-rc.2**

**Test 1**:

- adding remote node
```
curl -X POST ${DOJOT_HOST}/flows/v1/node -H "Authorization: Bearer ${JWT}" -H "content-type: application/json" -d '{"image": "dojot/kelvin-example:3.0.0-alpha2","id":"kelvin"}'
{"message":"ok"}
```
```
<13:44:22 24/11/2020> -- |nodeMngr -- DEBUG: Succeeded to add remote node with id kelvin
```
```
curl -X GET ${DOJOT_HOST}/flows/v1/node -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{
  "nodes": [
    {
      "id": "kelvin",
      "image": "dojot/kelvin-example:3.0.0-alpha2"
    }
  ]
}
```
```
mosquitto_pub -h ${HOST_MQTT} -p ${PORT_MQTT} -t ${tenant}:${id_device}/attrs -u ${tenant}:${id_device}  -m '{"temperatura":20}'
```
```
<13:46:04 24/11/2020> -- DEBUG: [executor] will handle node bce46601.d18af8:kelvin
<13:46:04 24/11/2020> -- |dispatcher -- DEBUG: Received reply: {"payload":[{"payload":{"metadata":{"tenant":"admin","timestamp":1606225564384},"event":"publish","data":{"attrs":{"temperatura":20,"protocol":"mqtt"},"id":"9d9643"}},"saida":{"temperatura":293.15}}],"requestId":"3fb0e1e1-40cf-4d67-8220-d29fbede4521"}
<13:46:04 24/11/2020> -- DEBUG: [executor] hop (kelvin) result: [{"payload":{"metadata":{"tenant":"admin","timestamp":1606225564384},"event":"publish","data":{"attrs":{"temperatura":20,"protocol":"mqtt"},"id":"9d9643"}},"saida":{"temperatura":293.15}}]
```
- restarting flowbroker
```
kubectl delete pod/flowbroker-548d4b849f-9cqzl
pod "flowbroker-548d4b849f-9cqzl" deleted
```
```
mosquitto_pub -h ${HOST_MQTT} -p ${PORT_MQTT} -t ${tenant}:${id_device}/attrs -u ${tenant}:${id_device}  -m '{"temperatura":30}'
```
```
<13:56:26 24/11/2020> -- DEBUG: [executor] will handle node bce46601.d18af8:kelvin
<13:56:26 24/11/2020> -- |dispatcher -- DEBUG: Received reply: {"payload":[{"payload":{"metadata":{"tenant":"admin","timestamp":1606226186012},"event":"publish","data":{"attrs":{"temperatura":30,"protocol":"mqtt"},"id":"9d9643"}},"saida":{"temperatura":303.15}}],"requestId":"c6a11589-e92e-47b5-9b52-4beac0935577"}
<13:56:26 24/11/2020> -- DEBUG: [executor] hop (kelvin) result: [{"payload":{"metadata":{"tenant":"admin","timestamp":1606226186012},"event":"publish","data":{"attrs":{"temperatura":30,"protocol":"mqtt"},"id":"9d9643"}},"saida":{"temperatura":303.15}}]
```
- restarting remote node
```
kubectl delete pod/flownode-adminkelvin-77659695-rfsv9
pod "flownode-adminkelvin-77659695-rfsv9" deleted
```
```
 curl -X GET ${DOJOT_HOST}/flows/v1/node -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{
  "nodes": [
    {
      "id": "kelvin",
      "image": "dojot/kelvin-example:3.0.0-alpha2"
    }
  ]
}
```
![Screenshot_36](https://user-images.githubusercontent.com/37310063/100105954-b12c7000-2e46-11eb-90fe-9ee6ed9832ed.png)

```
mosquitto_pub -h ${HOST_MQTT} -p ${PORT_MQTT} -t ${tenant}:${id_device}/attrs -u ${tenant}:${id_device}  -m '{"temperatura":10}'
```
```
<14:03:21 24/11/2020> -- DEBUG: [executor] will handle node bce46601.d18af8:kelvin
<14:03:21 24/11/2020> -- |dispatcher -- DEBUG: Received reply: {"payload":[{"payload":{"metadata":{"tenant":"admin","timestamp":1606226601270},"event":"publish","data":{"attrs":{"temperatura":10,"protocol":"mqtt"},"id":"9d9643"}},"saida":{"temperatura":283.15}}],"requestId":"66061092-f29f-4ea6-9b0b-0f5a900a8fd4"}
<14:03:21 24/11/2020> -- DEBUG: [executor] hop (kelvin) result: [{"payload":{"metadata":{"tenant":"admin","timestamp":1606226601270},"event":"publish","data":{"attrs":{"temperatura":10,"protocol":"mqtt"},"id":"9d9643"}},"saida":{"temperatura":283.15}}]
```
----
**Test 2**:

- delete kelvin
```
 curl -X DELETE ${DOJOT_HOST}/flows/v1/node/kelvin -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{
  "message": "ok"
}
```
```
curl -X GET ${DOJOT_HOST}/flows/v1/node -H "Authorization: Bearer ${JWT}" -H "content-type: application/json"
{
  "nodes": []
}
```
- adding kelvin
```
curl -X POST ${DOJOT_HOST}/flows/v1/node -H "Authorization: Bearer ${JWT}" -H "content-type: application/json" -d '{"image": "dojot/kelvin-example:3.0.0-alpha2","id":"kelvin"}'
{
  "message": "ok"
}
```
```
<14:09:19 24/11/2020> -- |nodeMngr -- DEBUG: Succeeded to add remote node with id kelvin
```
- restarting remote node
```
cpqd@k8s17-master-5:~$ kubectl delete pod/flownode-adminkelvin-77659695-vksch
pod "flownode-adminkelvin-77659695-vksch" deleted
```
```
mosquitto_pub -h ${HOST_MQTT} -p ${PORT_MQTT} -t ${tenant}:${id_device}/attrs -u ${tenant}:${id_device}  -m '{"temperatura":5}'
```
```
<14:13:52 24/11/2020> -- DEBUG: [executor] will handle node bce46601.d18af8:kelvin
<14:13:52 24/11/2020> -- |dispatcher -- DEBUG: Received reply: {"payload":[{"payload":{"metadata":{"tenant":"admin","timestamp":1606227231756},"event":"publish","data":{"attrs":{"temperatura":5,"protocol":"mqtt"},"id":"9d9643"}},"saida":{"temperatura":278.15}}],"requestId":"0face636-07f9-4cc6-b608-09bb05af689f"}
<14:13:52 24/11/2020> -- DEBUG: [executor] hop (kelvin) result: [{"payload":{"metadata":{"tenant":"admin","timestamp":1606227231756},"event":"publish","data":{"attrs":{"temperatura":5,"protocol":"mqtt"},"id":"9d9643"}},"saida":{"temperatura":278.15}}]
```
